src/decode.rs

```diff
@@ -2970,27 +2971,30 @@ unsafe fn decode_b_inner(
                     N_INTER_INTRA_PRED_MODES as usize - 1,
                 ) as u8;
                 let wedge_ctx = dav1d_wedge_ctx_lut[bs as usize] as c_int;
-                *b.interintra_type_mut() = INTER_INTRA_BLEND
-                    + rav1d_msac_decode_bool_adapt(
-                        &mut ts.msac,
-                        &mut ts.cdf.m.interintra_wedge[wedge_ctx as usize],
-                    ) as u8;
-                if b.interintra_type() == INTER_INTRA_WEDGE {
+                *b.interintra_type_mut() = if rav1d_msac_decode_bool_adapt(
+                    &mut ts.msac,
+                    &mut ts.cdf.m.interintra_wedge[wedge_ctx as usize],
+                ) {
+                    InterIntraType::Wedge
+                } else {
+                    InterIntraType::Blend
+                };
```

Comment:
Did you check if the `if` generates equally efficient code as the `+`?  `fn decode_b` is in the inner decoding loop, so it's likely quite performance sensitive.

Reply:
It does: https://godbolt.org/z/EoWGo61nv.

```rust
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum InterIntraType {
    None = 0,
    Blend = 1,
    Wedge = 2,
}

extern "C" {
    fn rav1d_msac_decode_bool_adapt() -> bool;
}

pub fn using_if() -> InterIntraType {
    if unsafe { rav1d_msac_decode_bool_adapt() } {
        InterIntraType::Wedge
    } else {
        InterIntraType::Blend
    }
}

pub fn using_add() -> u8 {
    InterIntraType::Blend as u8 + unsafe { rav1d_msac_decode_bool_adapt() } as u8
}
```

They both use `inc` (x86) and `cinc` (arm).

Reply:
the version that uses an `Option` is even better: the bool is just transmuted into a `Option<InterIntraType>` behind the scenes.